services/QuillLMS/app/controllers/snapshots_controller.rb

```diff
@@ -77,6 +60,32 @@ def options
     return render json: { error: 'user is not authorized for all specified schools' }, status: 403
   end
 
+  private def retrieve_cache_or_enqueue_worker(worker)
```

Comment:
Not totally sure about this abstraction as it sort of obfuscates what's going on in the actual actions, but it felt better than repeating all the code again.  Open to suggestions on approach tweaks.

Reply:
I think the abstraction makes sense.  We kind of do something similar with google classrooms [here](https://github.com/empirical-org/Empirical-Core/blob/develop/services/QuillLMS/app/controllers/teachers/classroom_manager_controller.rb#L202)

The timeframe makes the data model a little more complicated, but I don't see any easy way of simplifying it.